mujoco_warp/_src/smooth.py

```diff
@@ -1941,69 +1942,232 @@ def _spatial_site_tendon(
   site_bodyid: wp.array(dtype=int),
   wrap_objid: wp.array(dtype=int),
   tendon_site_pair_adr: wp.array(dtype=int),
-  wrap_site_adr: wp.array(dtype=int),
   wrap_site_pair_adr: wp.array(dtype=int),
+  wrap_pulley_scale: wp.array(dtype=float),
   # Data in:
   site_xpos_in: wp.array2d(dtype=wp.vec3),
   subtree_com_in: wp.array2d(dtype=wp.vec3),
   cdof_in: wp.array2d(dtype=wp.spatial_vector),
-  # In:
-  n_site_pair: int,
   # Data out:
   ten_length_out: wp.array2d(dtype=float),
   ten_J_out: wp.array3d(dtype=float),
-  wrap_obj_out: wp.array2d(dtype=wp.vec2i),
-  wrap_xpos_out: wp.array2d(dtype=wp.spatial_vector),
 ):
   worldid, elementid = wp.tid()
-  site_adr = wrap_site_adr[elementid]
 
-  site_xpos = site_xpos_in[worldid, wrap_objid[site_adr]]
+  # site pairs
+  site_pair_adr = wrap_site_pair_adr[elementid]
+  ten_adr = tendon_site_pair_adr[elementid]
+
+  # pulley scaling
+  pulley_scale = wrap_pulley_scale[site_pair_adr]
+
+  id0 = wrap_objid[site_pair_adr + 0]
+  id1 = wrap_objid[site_pair_adr + 1]
+
+  pnt0 = site_xpos_in[worldid, id0]
+  pnt1 = site_xpos_in[worldid, id1]
+  dif = pnt1 - pnt0
+  vec, length = math.normalize_with_norm(dif)
+  wp.atomic_add(ten_length_out[worldid], ten_adr, length * pulley_scale)
+
+  if length < MJ_MINVAL:
+    vec = wp.vec3(1.0, 0.0, 0.0)
+
+  body0 = site_bodyid[id0]
+  body1 = site_bodyid[id1]
+  if body0 != body1:
+    for i in range(nv):
```

Comment:
I could be wrong and the perf impact is not significant, but any time I see a `for i in range(nv)`, I would be inclined to add a TODO to fix it later

Reply:
added todos